### PR TITLE
evenstore proposal round 2

### LINKED ISSUE #12: Round 2 Design/Implementation
Make changes resulted from a discussion of the first review of the proposal.

### PR DESCRIPTION
**Pull request by jsign**
This is a continuation of #6 
Closes #12

After the first round of discussion, there are multiple changes:
* `.Add` was changed to `.Create`
* `ID`s of instances are typed with `EntityID`
* the `ID` attribute in schemas is enforced when registering the schema in the model
* when `.Create` is called, the `Model` automatically generates a new `EntityID` for the instance, to make instance creation more friendly to the client.
* `.Read` and `.Update` tnx of the model were renamed to `.ReadTxn` and `.WriteTxn` for clarity (semantics are the same)
*  introduced the concept of `EventCreator` which is the component that generates events for `Dispatcher`, and also has the knowledge to make the `Reduce` work. (some more comments about this later)
* quite a big refactor in general, more confident now that is easier to understand/read the code

I'll add below some further comments on the code to help to understand and maybe ask for some opinions.

### CODE REVIEW

File: eventcreator.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+type ActionType int
+
+const (
+	Create ActionType = iota
+	Save
+	Delete
+)
+
+// Action is a operation done in the model
+type Action struct {
```
Comment (jsign) [inline eventcreator.go:16]: The idea of `Action` is to represents actions that happened in a model transaction. When the transaction is `Commit`ed, the list of `Action`s that happened are delegated to the `EventCreator` to generate proper events to later be dispatched to the `Dispatcher`. (I'll reference this point in code in another comment).

File: jsonpatcher/jsonpatcher.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
+package jsonpatcher
```
Comment (jsign) [inline jsonpatcher/jsonpatcher.go:1]: A `jsonpatcher` is an implementation of the `EventCreator` interface.
It basically translates `Action`s to `Events`. In this case, is an implementation that uses json-patches to generate payloads for the Events.  Later when new `Event`s are received from the `Dispatcher`, it unmarshals the payloads again to json-patches which are applied to the underlying `datastore`.

At this point something became somewhat apparent. Since the `EventCreator` is encoding the changes into the generated `Events`, it's the correct party to interpret new `Reduce`d events... since its the component that understand how things are encoded in and out of `Events`.

Despite being a reasonable realization... it still makes me wonder if the name `EventCreator` is the correct thing... since it seems more like a `EventCodec` (since does encoding and decoding of `Action`s to `Event`s and viceversa). For the moment I kept the original name from the WP, but I'd like some opinions if this makes sense.

File: jsonpatcher/jsonpatcher.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+type patcher struct {
+}
+
+var _ es.EventCreator = (*patcher)(nil)
+
+func New() es.EventCreator {
+	return &patcher{}
+}
+
+func (m *patcher) Create(actions []es.Action) ([]es.Event, error) {
```
Comment (jsign) [inline jsonpatcher/jsonpatcher.go:47]: Here is where a set of `Action`s done in a `Model` txn, is translated to `Events`. It basically generates a json with the operation that was done, and:
* the instance full json if was a `.Create`
* the json-diff of the original instance and the updated instance if was a `.Save`
* nothing in payload if was a `.Delete` since only the `EntityID` is sufficient.

File: jsonpatcher/jsonpatcher.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+			Timestamp: time.Now(),
+			ID:        actions[i].EntityID,
+			TypeName:  actions[i].EntityType,
+			Patch:     eventPayload,
+		}
+	}
+	return events, nil
+}
+
+func (p *patcher) Reduce(e es.Event, datastore ds.Datastore, baseKey ds.Key) error {
```
Comment (jsign) [inline jsonpatcher/jsonpatcher.go:75]: Here's basically the inverse of `jsonpatcher.Create()`.  `jsonpatcher.Create()` and `jsonpatcher.Reduce()` are the back and forth flow of changes in the model. Could be understood as `encode` and `decode` respectively.

File: store/README.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
+# API Rationale
```
Comment (jsign) [inline store/README.md:1]: Here's an updated doc with some explanation of how the current API behaves. Might help to a more high-level feeling of how it looks now.

File: store/model.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+type Txn struct {
+	model     *Model
+	discarded bool
+	commited  bool
+	readonly  bool
+
+	actions []es.Action
+}
+
+type SaveOp struct {
```
Comment (jsign) [inline store/model.go:113]: Ignore this, since it will be deleted! Not want to invalidate this review at this point :smile:

File: store/model.go [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+	}
+
+	a := es.Action{
+		Type:       es.Create,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    new,
+	}
+	t.actions = append(t.actions, a)
```
Comment (jsign) [inline store/model.go:142]: Every `Create`, `Save` and `Delete` generates an `Action` in the Txn state, that will be processed on `Commit()`

File: store/model.go [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+	beforeBytes, err := t.model.datastore.Get(key)
+	if err == ds.ErrNotFound {
+		return errCantSaveNonExistentInstance
+	}
+	if err != nil {
+		return err
+	}
+
+	before := reflect.New(t.model.valueType.Elem()).Interface()
+	err = json.Unmarshal(beforeBytes, before)
```
Comment (jsign) [inline store/model.go:163]: Interesting `reflect` magic. Since `Txn` is a generic transaction for some particular instance `Type`, this code allows generating a default zero-value instance of a particular Go type, since we need it for unmarshaling the JSON representation that is in the store in `[]byte`.

File: store/model.go [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+	}
+
+	events, err := t.model.eventcreator.Create(t.actions)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range events {
+		if err := t.model.dispatcher.Dispatch(e); err != nil {
+			return err // ToDo/Note: Important to document the implications of a partial dispatch
```
Comment (jsign) [inline store/model.go:235]: Important comment here... I'd like to think about what should happen here. Not really trivial.

File: store/model.go [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+func (m *Model) Reduce(event es.Event) error {
+	// ToDo: distinguish local and remote events for proper locking
+	log.Debugf("reducer %s start", m.schema.Ref)
+	if event.Type() != m.schema.Ref {
+		log.Debugf("ignoring event from uninteresting type")
+		return nil
+	}
+
+	return m.eventcreator.Reduce(event, m.datastore, m.dsKey)
```
Comment (jsign) [inline store/model.go:249]: This is something that feels a bit weird, since the `Model.Reduce` is mostly delegating the work to the `EventCreator`. Since a priori the `Model` can be used with a JSON-patcher, or JSON-CRDT or who knows which other implementation... so the `Model` doesn't really know what was encoded in the `Event` payload.

I went back and forth with making the `EventCreator` the real `Reducer` interface implementation... but I had some conflicting feelings since that would mean that the `EventCreator` would be the owner of the `datastore`... 

Not fully unconvinced, but if it makes sense to anyone to have some opinion, it would be appreciated. :sweat_smile:

File: store/model.go [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+func (t *Txn) Discard() {
+	t.discarded = true
+}
+
+func getEntityID(t interface{}) es.EntityID {
```
Comment (jsign) [inline store/model.go:256]: Some magic to extract the `EntityID` from an `interface{}` which represents an instance of a schema. Maybe be simplified; more important things pending. Works correctly.

File: store/model.go [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+	for _, e := range events {
+		if err := t.model.dispatcher.Dispatch(e); err != nil {
+			return err // ToDo/Note: Important to document the implications of a partial dispatch
+		}
+	}
+	return nil
+}
+
+func (m *Model) Reduce(event es.Event) error {
+	// ToDo: distinguish local and remote events for proper locking
```
Comment (jsign) [inline store/model.go:242]: Super important comment here that I'd like some opinions.

On every Txn start, a mutex is locked for the Model, in order to serialize changes and let the txn logic be sure that not other things are happening until after commits. Simple.

But we have a tricky situation. When the `Commit` is done, the generated `Events` are dispatcher to the `Dispatcher`, which indirectly calls the model `Reducer` which really applies the changes to the datastore.

The problem is that a call to `Reduce` here, can't be distinguished from a local committed transaction, or a remote event that came from the network... so adding in the quoted line a `lock.Lock()` can't be done. If the call to the `Reduce` is from a txn `Commit`, then the lock is already locked. 

The first thing that cames to my mind is to change the `Reduce` parameters to add an extra `local bool` parameter. If the `Dispatcher` sent `true` it means that we don't need to `Lock()` the mutex since implicitly means that is already locked. If it is `false`, it means that the `Event` sent from the `Dispatcher` isn't from a local txn but received from the network, so it's safe to `Lock()` to not interfere with local action.

Hope this explanation makes sense... if not, just ask :smile:

File: store/store.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	if !isValidModel(defaultInstance) {
+		return nil, ErrInvalidModel
+	}
+
+	eventcreator := jsonpatcher.New()
+	m := NewModel(name, defaultInstance, s.datastore, s.dispatcher, eventcreator)
+	s.models[m.valueType] = m
+	s.dispatcher.Register(m) // ToDo: find good place for reg token, with proper unregistering
```
Comment (jsign) [inline store/store.go:50]: I'm wondering if a `Close()` method should exist in a `Store` that will unregister all the reducers from the dispatcher. Because `dispatcher.Register(..)` returns a registration token that allows the reducer to be unregistered. When this should happen? my best guess is in some `Close()` call to the `Store` that should exist?

**Review by jsign — COMMENTED** [review]
Lots of comments to help understand and ask for opinions!

File: store/model.go (hunk 12 above)
Comment (jsign) [reply store/model.go:242]: Flash insight: The solution is actually more simple (since above one is plain incorrect since can have deadlocks).

Write transactions should acquire a write lock of the `Dispatcher`. Read transactions should acquire a read lock of the `Dispatcher`. Saying it differently, a txn is `Dispatcher`-wise, not model wise, since every model `Commit` can affect other models too (since `Dispatcher` dispatch events to all registered reducer and thus different models).

TL;DR: A `WriteTxn` is a stop-the-world thing to let the logic do things safely. Only one `WriteTxn` can exist at anytime in all models registered in the `Dispatcher`. The good thing about having `ReadTxn` is that multiple ones can be running at the same time. Any *remote* events dispatched remotely (because new events from logs where discovered) would be like a `WriteTxn`.

Just writing this because it popped out in my mind now.

Edit: All this considering that ordering *may* be important on event delivery, at least for local changes. If that isn't the case, we can just let commutativity property solve concurrency.

File: jsonpatcher/jsonpatcher.go (hunk 2 above)
Comment (sanderpick) [reply jsonpatcher/jsonpatcher.go:1]: Makes sense to me. `EventCodec` has a nice ring to it.

File: event.go [hunk 14]
```diff
@@ -4,14 +4,30 @@ import (
...
+	"github.com/google/uuid"
+)
+
+const (
+	EmptyEntityID = EntityID("")
 )
 
+type EntityID string
+
+func NewEntityID() EntityID {
```
Comment (sanderpick) [inline event.go:17]: Great!

File: ignore_view.go [hunk 15]
```diff
@@ -0,0 +1,61 @@
+package eventstore
+
+// import (
```
Comment (sanderpick) [inline ignore_view.go:3]: Is it safe to remove these ignore_* files?

File: store/model.go (hunk 8 above)
Comment (sanderpick) [reply store/model.go:163]: Nice, that's handy

File: store/model.go (hunk 9 above)
Comment (sanderpick) [reply store/model.go:235]: I suppose ideally it would be trivial but that's tricky! At least we're stopping at the error for now. There may be some inspiration to take from Realm.

File: store/store.go (hunk 13 above)
Comment (sanderpick) [reply store/store.go:50]: Agreed, de-register on Close makes sense. Do we need to de-register tokens? I can't imagine a case where they'd be used.

File: store/model.go (hunk 12 above)
Comment (sanderpick) [reply store/model.go:242]: Lots of wires in here! 😅 I like the consideration of keeping order where possible.

File: store/model.go (hunk 10 above)
Comment (sanderpick) [reply store/model.go:249]: Mapping components to logical concern makes sense to me even if that means some are small, though I don't have an intuition here.

**Review by sanderpick — DISMISSED** [review]
I left some comments / questions but all the changes look great! Nice tests to boot.

File: dispatcher.go [hunk 16]
```diff
@@ -13,6 +13,10 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
+type Reducer interface {
+	Reduce(event Event) error
+}
+
 // @todo: Should we also support a `Transformer` to actually fetch raw event data as part of a pipeline?
```
Comment (carsonfarmer) [inline dispatcher.go:20]: might as well delete this next time you're in there?

File: dispatcher.go [hunk 17]
```diff
@@ -73,7 +77,7 @@ func (d *Dispatcher) Dispatch(event Event) error {
 	defer d.lock.Unlock()
 	// Key format: <timestamp>/<entity-id>/<type>
 	// @todo: This is up for debate, its a 'fake' Event struct right now anyway
-	key := datastore.NewKey(string(event.Time())).ChildString(event.EntityID()).ChildString(event.Type())
+	key := datastore.NewKey(string(event.Time())).ChildString(event.EntityID().String()).ChildString(event.Type())
```
Comment (carsonfarmer) [inline dispatcher.go:80]: Do you still think having time first is a good idea? I was starting to wonder if prefix searching via entity it would be useful. Then you'd want something like `<entity id>/<time>/<type>/<uid>`?

File: eventcreator.go (hunk 1 above)
Comment (carsonfarmer) [reply eventcreator.go:16]: Very nice.

File: eventcreator.go [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+	// Type of the action
+	Type ActionType
+	// EntityID of the instance in action
+	EntityID EntityID
+	// EntityType of the instance in action
+	EntityType string
+	// Before is the instance before operation was done
```
Comment (carsonfarmer) [inline eventcreator.go:23]: Names are off here now... Before -> Previous, After -> Current

File: eventcreator.go [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+	Previous interface{}
+	// After is the instance after
+	Current interface{}
+}
+
+type EventCreator interface {
```
Comment (carsonfarmer) [inline eventcreator.go:29]: Nice

File: ignore_view.go (hunk 15 above)
Comment (carsonfarmer) [reply ignore_view.go:3]: Yes, should have been deleted already... I guys that PR never was merged. Delete away.

File: ignore_view_test.go [hunk 20]
```diff
@@ -0,0 +1,79 @@
+package eventstore
```
Comment (carsonfarmer) [inline ignore_view_test.go:1]: Delete

File: jsonpatcher/jsonpatcher.go (hunk 2 above)
Comment (carsonfarmer) [reply jsonpatcher/jsonpatcher.go:1]: Yeh love this layer for the reducing functionality, makes great sense. I also like EventCodec... if that feels 'right', I'd say go with it. The only reason to keep EventCreator would be to stick with existing Flux pattern naming conventions (which we've deviated from enough that I don't think it matters).

File: jsonpatcher/jsonpatcher.go (hunk 3 above)
Comment (carsonfarmer) [reply jsonpatcher/jsonpatcher.go:47]: Love it.

File: jsonpatcher/jsonpatcher.go (hunk 4 above)
Comment (carsonfarmer) [reply jsonpatcher/jsonpatcher.go:75]: If you go with EventCodec, you might want to consider _using_ those names?

Comment (carsonfarmer) [reply jsonpatcher/jsonpatcher.go:75]: This is a really nice clear implementation here... really like it. Easy to follow.

File: store/README.md (hunk 5 above)
Comment (carsonfarmer) [reply store/README.md:1]: LOVE LOVE LOVE

File: store/model.go [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+
+	ErrNotFound   = errors.New("instance not found")
+	ErrReadonlyTx = errors.New("read only transaction")
+
+	errAlreadyDiscardedCommitedTxn = errors.New("can't commit discarded/commited txn")
+	errCantCreateExistingInstance  = errors.New("can't create already existing instance")
+	errCantSaveNonExistentInstance = errors.New("can't save unkown instance")
+)
+
+type Model struct {
```
Comment (carsonfarmer) [inline store/model.go:25]: Niiiice

File: store/model.go [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+		return err
+	}
+	if !exists {
+		return ErrNotFound
+	}
+	a := es.Action{
+		Type:       es.Delete,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
```
Comment (carsonfarmer) [inline store/model.go:195]: Is it at all useful to keep Previous here, in case we want to 'undo' something, or rollback a transaction? Or is it easier to just leave this up to the underlying store?

File: store/model.go (hunk 9 above)
Comment (carsonfarmer) [reply store/model.go:235]: Yeh, ideally, you'd rollback any transactions that resulted from the dispatch... this is something I really struggled with previously, and never had a good solution. I think I convinced myself that partial dispatches were ok... but not so sure about that? Also, I think we could leave some of these decisions up to the caller?

File: store/model.go (hunk 12 above)
Comment (carsonfarmer) [reply store/model.go:242]: Agreed, if we can preserve arrival ordering, we're less reliant on CRDT structures down the line if we don't want them. I think your flash insight is on the money for sure. Seems reasonable. Let's go with that for now, and see how it plays out?

File: store/model.go (hunk 10 above)
Comment (carsonfarmer) [reply store/model.go:249]: You've convinced _me_ that having the EventCreator be the Reducer... seems like the right abstraction to me from the outside?

File: store/store.go (hunk 13 above)
Comment (carsonfarmer) [reply store/store.go:50]: Yeh, we inherited the tokens from the work I had initially done, which was copied from Flux... but we probs don't need them here, because its never going to be 'dynamic' like that.

**Review by carsonfarmer — DISMISSED** [review]
Awesome stuff @jsign, this feels like an API I'd enjoy using!

File: store/model.go (hunk 9 above)
Comment (jsign) [reply store/model.go:235]: Created #15 to keep track of this topic since is quite complex. I tried to write there a summary of the problem and some ideas that are now in my mind, feel free to add more points there if something extra is fresh now in yours minds.

File: store/model.go (hunk 12 above)
Comment (jsign) [reply store/model.go:242]: Right, so I'll inclined towards a safe in-order approach, and safeguard all models state registered in the `Dispatcher`. If appears to be a problem in the future, we can have another round of dicussion!

File: dispatcher.go (hunk 17 above)
Comment (jsign) [reply dispatcher.go:80]: TBH, didn't look much more in the `dispatcher.go` after my last review in your PR.
I'm not sure yet when we're going to query the _raw_ events from the Dispatcher... I guess having the answer to that question would make that decision more clear. 

Do you have a clear situation in mind in which raw events would be queried?

Comment (jsign) [reply dispatcher.go:80]: Left in #16 to keep a record of this topic.

File: eventcreator.go (hunk 18 above)
Comment (jsign) [reply eventcreator.go:23]: Oops, thanks!

File: store/model.go (hunk 22 above)
Comment (jsign) [reply store/model.go:195]: Oh, that's sounds interesting. I added it to #15 to keep remember that.

File: store/store.go (hunk 13 above)
Comment (jsign) [reply store/store.go:50]: Not sure about it then... just kept it to not make more unilateral decisions. Maybe I'll remove the concept of the token and let the future give the hint if it is needed.

**Comment by jsign** [comment]
Thanks, @sanderpick @carsonfarmer  for the reviews.

I summarized some changes to do now (in the next PR that will arrive soon) in #14. Other more complex or unknowns are in #15 and #16.
